Mathlib/MeasureTheory/SetSemiring.lean

```diff
@@ -0,0 +1,279 @@
+/-
+Copyright (c) 2023 Rémy Degenne. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Rémy Degenne, Peter Pfaffelhuber
+-/
+import Mathlib.Data.Set.Pairwise.Lattice
+import Mathlib.Data.Real.ENNReal
+import Mathlib.MeasureTheory.PiSystem
+
+/-! # Semirings of sets
+
+A semi-ring of sets `C` (in the sense of measure theory) is a family of sets containing `∅`,
+stable by intersection and such that for all `s, t ∈ C`, `t \ s` is equal to a disjoint union of
+finitely many sets in `C`. Note that a semi-ring of sets may not contain unions.
+
+An important example of a semi-ring of sets is intervals in `ℝ`. The intersection of two intervals
+is an interval (possibly empty). The union of two intervals may not be an interval.
+The set difference of two intervals may not be an interval, but it will be a disjoint union of
+two intervals.
+
+## Main definitions
+
+* `MeasureTheory.IsSetSemiring C`: property of being a semi-ring of sets.
+* `MeasureTheory.IsSetSemiring.diffFinset hs ht`: for `s, t` in a semi-ring `C`
+  (with `hC : IsSetSemiring C`) with `hs : s ∈ C`, `ht : t ∈ C`, this is a `Finset` of
+  pairwise disjoint sets such that `s \ t = ⋃₀ hC.diffFinset hs ht`.
+* `MeasureTheory.IsSetSemiring.diffFinset₀ hs hI`: for `hs : s ∈ C` and a finset `I` of sets in `C`
+  (with `hI : ↑I ⊆ C`), this is a `Finset` of pairwise disjoint sets such that
+  `s \ ⋃₀ I = ⋃₀ hC.diffFinset₀ hs hI`.
+
+## Main statements
+
+* `MeasureTheory.IsSetSemiring.exists_disjoint_finset_diff_eq`: the existence of the `Finset` given
+  by the definition `IsSetSemiring.diffFinset₀` (see above).
+
+-/
+
+open Finset Set
+
+open scoped ENNReal BigOperators
+
+namespace MeasureTheory
+
+variable {α : Type*} {C : Set (Set α)} {s t : Set α}
+
+/-- A semi-ring of sets `C` is a family of sets containing `∅`, stable by intersection and such that
+for all `s, t ∈ C`, `s \ t` is equal to a disjoint union of finitely many sets in `C`. -/
+structure IsSetSemiring (C : Set (Set α)) : Prop where
+  empty_mem : ∅ ∈ C
+  inter_mem : ∀ s ∈ C, ∀ t ∈ C, s ∩ t ∈ C
+  diff_eq_Union' : ∀ s ∈ C, ∀ t ∈ C,
+    ∃ I : Finset (Set α), ↑I ⊆ C ∧ PairwiseDisjoint (I : Set (Set α)) id ∧ s \ t = ⋃₀ I
+
+namespace IsSetSemiring
+
+lemma isPiSystem (hC : IsSetSemiring C) : IsPiSystem C := fun s hs t ht _ ↦ hC.inter_mem s hs t ht
+
+section diffFinset
+
+/-- In a semi-ring of sets `C`, for all `s, t ∈ C`, `s \ t` is equal to a disjoint union of finitely
+many sets in `C`. This definitions gives a finset of sets that satisfies that equality.
+
+We remove the empty set to ensure that `t ∉ hC.diffFinset hs ht` even if `t = ∅`. -/
+noncomputable def diffFinset (hC : IsSetSemiring C) (hs : s ∈ C) (ht : t ∈ C)
+    [DecidableEq (Set α)] :
```

Comment:
I'd drop the decidability assumption, and use instead 
```
open Classical in
/-- ... -/
noncomputable def ...
```
This means that the first few proofs below setting up the API would have to start with `classical`, but once this is done you get can rid of the decidability assumption in all downstream applications.